source/common/network/apple_dns_impl.cc

```diff
@@ -0,0 +1,325 @@
+    }
+  }
+
+  // Purge the contents so no one tries to delete them again.
+  queries_with_pending_cb_.clear();
+
+  if (with_error) {
+    // The main sd ref is destroyed here because a callback with an error is good indication that
+    // the connection to the DNS server is faulty and needs to be torn down.
+    //
+    // Deallocation of the MainSdRef __has__ to happen __after__ flushing queries. Flushing queries
+    // de-allocates individual refs, so deallocating the main ref ahead would cause deallocation of
+    // invalid individual refs per dns_sd.h
+    deallocateMainSdRef();
+    initializeMainSdRef();
+  }
+}
+
+AppleDnsResolverImpl::PendingResolution::~PendingResolution() {
+  ENVOY_LOG(debug, "Destroying PendingResolution for {}", dns_name_);
+  DNSServiceRefDeallocate(individual_sd_ref_);
+}
+
+void AppleDnsResolverImpl::PendingResolution::cancel() {
+  ENVOY_LOG(debug, "Cancelling PendingResolution for {}", dns_name_);
+  ASSERT(owned_);
+  if (pending_cb_) {
+    /* (taken and edited from dns_sd.h)
+     * Canceling operations and kDNSServiceFlagsMoreComing
+     * Whenever you cancel any operation for which you had deferred [resolution]
+     * because of a kDNSServiceFlagsMoreComing flag, you should [flush]. This is because, after
+     * cancelling the operation, you can no longer wait for a callback *without* MoreComing set, to
+     * tell you [to flush] (the operation has been canceled, so there will be no more callbacks).
+     *
+     * [FURTHER] An implication of the collective
+     * kDNSServiceFlagsMoreComing flag for shared connections is that this
+     * guideline applies more broadly -- any time you cancel an operation on
+     * a shared connection, you should perform all deferred updates for all
+     * operations sharing that connection. This is because the MoreComing flag
+     * might have been referring to events coming for the operation you canceled,
+     * which will now not be coming because the operation has been canceled.
+     */
+    // First, get rid of the current query, because if it is canceled, its callback should not be
+    // executed during the subsequent flush.
+    parent_.removePendingQuery(this);
+    // Then, flush all other queries.
+    parent_.flushPendingQueries(false /* with_error */);
+  }
+  // Because the query is self-owned, delete now.
+  delete this;
+}
+
+void AppleDnsResolverImpl::PendingResolution::onDNSServiceGetAddrInfoReply(
+    DNSServiceFlags flags, uint32_t interface_index, DNSServiceErrorType error_code,
+    const char* hostname, const struct sockaddr* address, uint32_t ttl) {
+  ENVOY_LOG(debug,
+            "DNS for {} resolved with: flags={}[MoreComing={}, Add={}], interface_index={}, "
+            "error_code={}, hostname={}",
+            dns_name_, flags, flags & kDNSServiceFlagsMoreComing ? "yes" : "no",
+            flags & kDNSServiceFlagsAdd ? "yes" : "no", interface_index, error_code, hostname);
+  ASSERT(interface_index == 0);
```

Comment:
How certain are you of some of these ASSERTs on state that comes from the NS libs?

Reply:
As certain as I can be from what they write in the comments in dns_sd.h 😅